ovn-org/ovn-kubernetes
# netpol: Changes to support network policy for host network traffic

This commit adds support for enabling network policy on host
network traffic pods such as the router pod in OpenShift.

Currently, if you have a default deny network policy on a namespace,
services within that namespace are not reachable via host network
using external routes. This feature will make that possible by allowing
users to create a separate network policy with a host network namespace selector.
   
The expectation is that the namespace you want to use to classify host network
traffic to is pre-created with the desired labels. These labels will then
be used as match fields for the network policy object's ingress selector.
It is also necessary that the users pass the host network namespace
as a command line parameter for ovnkube when it starts.

**How to test this feature:**
- Start ovnkube with --host-network-namespace parameter set to `ovn-host-network` namespace.
- Create the `ovn-host-network` namespace. This namespace will be used to classify host network traffic with appropriate labels.
```
apiVersion: v1
kind: Namespace
metadata:
  # NOTE: this namespace is the portal for host network traffic
  name: ovn-host-network
  labels:
    policy-group.k8s.ovn.org/host-network: ""
```
- Create a target namespace that can be used to test network policy 
```
apiVersion: v1
kind: Namespace
metadata:
  # NOTE: this namespace is the portal for host network traffic
  name: test
```
- Create a service in the `test` namespace that needs to be accessible from outside. 
For instance create an nginx service that runs on port 80:
```
kubectl create deployment nginx
kubectl expose deployment nginx --port=80
```
- Verify that this service is reachable from outside.
- Create a default-deny policy in the `test` namespace.
```
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: default-deny-ingress
spec:
  podSelector: {}
  policyTypes:
  - Ingress
```
- Verify that this service is now not reachable from outside.
- Create a network policy that matches on the labels on the `ovn-host-network` namespace and uses the 
pod selector to target the nginx service created above.
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: access-nginx
spec:
  podSelector:
    matchLabels:
      app: nginx
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          policy-group.k8s.ovn.org/host-network: ""
```
- Verify that the service is reachable again.

Signed-off-by: Aniket Bhat <anbhat@redhat.com>
